File: day09/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
)

const (
	puzzleInput = "input.txt"
)

type (
	Grid struct {
		grid  [][]byte
		h     int
		w     int
		basin [][]byte
	}
)

func NewGrid(grid [][]byte) *Grid {
	h := len(grid)
	w := len(grid[0])
	basin := make([][]byte, h)
	for i := range basin {
		basin[i] = make([]byte, w)
	}
	return &Grid{
		grid:  grid,
		h:     h,
		w:     w,
		basin: basin,
	}
}

func (g *Grid) outBounds(x, y int) bool {
	return x < 0 || y < 0 || x >= g.w || y >= g.h
}

func (g *Grid) inBounds(x, y int) bool {
	return !g.outBounds(x, y)
}

func (g *Grid) isLow(x, y int) bool {
	v := g.grid[y][x]
	if g.inBounds(x-1, y) && g.grid[y][x-1] <= v {
		return false
	}
	if g.inBounds(x, y-1) && g.grid[y-1][x] <= v {
		return false
	}
	if g.inBounds(x+1, y) && g.grid[y][x+1] <= v {
		return false
	}
	if g.inBounds(x, y+1) && g.grid[y+1][x] <= v {
		return false
	}
	return true
}

func (g *Grid) markBasin(k byte, x, y int) int {
	if !g.inBounds(x, y) {
		return 0
	}
	if g.grid[y][x] == '9' {
		return 0
	}
	if g.basin[y][x] != 0 {
		return 0
	}
	g.basin[y][x] = k
	return 1 + g.markBasin(k, x-1, y) + g.markBasin(k, x, y-1) + g.markBasin(k, x+1, y) + g.markBasin(k, x, y+1)
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var rows [][]byte

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		rows = append(rows, []byte(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	grid := NewGrid(rows)

	count := 0
	var basinnum byte = 1
	var sizes []int
	for r, i := range grid.grid {
		for c := range i {
			if grid.isLow(c, r) {
				count += int(grid.grid[r][c]-'0') + 1
				size := grid.markBasin(basinnum, c, r)
				sizes = append(sizes, size)
				basinnum++
			}
		}
	}
	fmt.Println("Part 1:", count)
	if len(sizes) >= 3 {
		sort.Ints(sizes)
		l := len(sizes)
		fmt.Println("Part 2:", sizes[l-1]*sizes[l-2]*sizes[l-3])
	}
}

File: day09/src/main.rs
use std::cmp::Reverse;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct Grid {
    rows: Vec<Vec<u8>>,
    h: usize,
    w: usize,
    basin: Vec<Vec<bool>>,
}

impl Grid {
    fn new(rows: Vec<Vec<u8>>) -> Self {
        let h = rows.len();
        let w = rows[0].len();
        Self {
            rows,
            h,
            w,
            basin: (0..h).into_iter().map(|_| vec![false; w]).collect(),
        }
    }

    fn out_bounds(&self, x: usize, y: usize) -> bool {
        x >= self.w || y >= self.h
    }

    fn in_bounds(&self, x: usize, y: usize) -> bool {
        !self.out_bounds(x, y)
    }

    fn is_low(&self, x: usize, y: usize) -> bool {
        let v = self.rows[y][x];
        if self.in_bounds(x.wrapping_sub(1), y) && self.rows[y][x.wrapping_sub(1)] <= v {
            return false;
        }
        if self.in_bounds(x, y.wrapping_sub(1)) && self.rows[y.wrapping_sub(1)][x] <= v {
            return false;
        }
        if self.in_bounds(x + 1, y) && self.rows[y][x + 1] <= v {
            return false;
        }
        if self.in_bounds(x, y + 1) && self.rows[y + 1][x] <= v {
            return false;
        }
        true
    }

    fn mark_basin(&mut self, x: usize, y: usize) -> i32 {
        if !self.in_bounds(x, y) {
            return 0;
        }
        if self.rows[y][x] == b'9' {
            return 0;
        }
        if self.basin[y][x] {
            return 0;
        }
        self.basin[y][x] = true;
        1 + self.mark_basin(x.wrapping_sub(1), y)
            + self.mark_basin(x, y.wrapping_sub(1))
            + self.mark_basin(x + 1, y)
            + self.mark_basin(x, y + 1)
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut grid = Grid::new(
        reader
            .lines()
            .flat_map(|i| i)
            .map(|i| Vec::from(i.as_bytes()))
            .collect(),
    );
    let w = grid.w;
    let mut sizes = (0..grid.h)
        .into_iter()
        .flat_map(|r| (0..w).into_iter().map(move |c| (r, c)))
        .flat_map(|(r, c)| {
            if grid.is_low(c, r) {
                Some((
                    (grid.rows[r][c] as i32) - (b'0' as i32) + 1,
                    grid.mark_basin(c, r),
                ))
            } else {
                None
            }
        })
        .collect::<Vec<_>>();
    sizes.sort_by_key(|&(_, a)| Reverse(a));
    println!("Part 1: {}", sizes.iter().map(|&(a, _)| a).sum::<i32>());
    println!(
        "Part 2: {}",
        sizes.iter().take(3).map(|&(_, a)| a).product::<i32>()
    );
    Ok(())
}
